src/parser/ast_node/ast_node.rs

```diff
@@ -0,0 +1,47 @@
+// #[derive(Debug)]
+use super::ast_node_concat::AstNodeConcat;
+use super::ast_node_group::AstNodeGroup;
+use super::ast_node_literal::AstNodeLiteral;
+use super::ast_node_optional::AstNodeOptional;
+use super::ast_node_plus::AstNodePlus;
+use super::ast_node_star::AstNodeStar;
+use super::ast_node_union::AstNodeUnion;
+
+pub(crate) enum AstNode {
+    Literal(AstNodeLiteral),   // Single character literal
+    Concat(AstNodeConcat),     // Concatenation of two expressions
+    Union(AstNodeUnion),       // Union of two expressions
+    Star(AstNodeStar),         // Kleene Star (zero or more)
+    Plus(AstNodePlus),         // One or more
+    Optional(AstNodeOptional), // Zero or one (optional)
+    Group(AstNodeGroup),       // Capturing group
+}
+
+impl PartialEq for AstNode {
+    fn eq(&self, other: &Self) -> bool {
+        match (self, other) {
+            (AstNode::Literal(l1), AstNode::Literal(l2)) => l1 == l2,
+            (AstNode::Concat(c1), AstNode::Concat(c2)) => c1 == c2,
+            (AstNode::Union(u1), AstNode::Union(u2)) => u1 == u2,
+            (AstNode::Star(s1), AstNode::Star(s2)) => s1 == s2,
+            (AstNode::Plus(p1), AstNode::Plus(p2)) => p1 == p2,
+            (AstNode::Optional(o1), AstNode::Optional(o2)) => o1 == o2,
+            (AstNode::Group(g1), AstNode::Group(g2)) => g1 == g2,
```

Comment:
```suggestion
            (AstNode::Literal(lhs), AstNode::Literal(rhs)) => lhs == rhs,
            (AstNode::Concat(lhs), AstNode::Concat(rhs)) => lhs == rhs,
            (AstNode::Union(lhs), AstNode::Union(rhs)) => lhs == rhs,
            (AstNode::Star(lhs), AstNode::Star(rhs)) => lhs == rhs,
            (AstNode::Plus(lhs), AstNode::Plus(rhs)) => lhs == rhs,
            (AstNode::Optional(lhs), AstNode::Optional(rhs)) => lhs == rhs,
            (AstNode::Group(lhs), AstNode::Group(rhs)) => lhs == rhs,
```
Let's use `lhs` and `rhs` for clarifications